### PR TITLE
Solr search

### PR DESCRIPTION
**Pull request by grandpaslab**
Shim to point search at a Solr instance.

### CODE REVIEW

**Comment by kemitche** [comment]
High level comment: Can you refactor the `SEARCH_TYPE` toggle to follow our "provider" pattern? You can use the media_provider as an example; see line ~430 of `app_globals.py` and the code under `r2/r2/lib/providers/media` for examples on how that works, but the short of it is you have a base class that defines the interface, and subclasses that define required config and the implementation of the provider.

File: r2/r2/lib/solrsearch.py [hunk 1]
```diff
@@ -0,0 +1,564 @@
...
+from lxml import etree
+from r2.lib import amqp, filters
+from r2.lib.db.sorts import epoch_seconds
+import r2.lib.utils as r2utils
+from r2.models import (Account, Subreddit, All, DefaultSR,
+                       MultiReddit, DomainSR, Friends, ModContribSR,
+                       FakeSubreddit)
+from pylons import g, c
+
+from r2.lib.cloudsearch import *
```
Comment (kemitche) [inline r2/r2/lib/solrsearch.py]: Please don't use `import *`; it can cause subtle import bugs (such as shadowing the `datetime` class with the `datetime` module).

File: r2/r2/lib/solrsearch.py [hunk 2]
```diff
@@ -0,0 +1,564 @@
...
+    def _get_sr_restriction(sr):
+        '''Return a solr-appropriate query string that restricts
+        results to only contain results from self.sr
+        
+        '''
+        bq = []
+        if (not sr) or sr == All or isinstance(sr, DefaultSR):
+            return None
+        elif isinstance(sr, MultiReddit):
+            bq = ["(or"]
```
Comment (kemitche) [inline r2/r2/lib/solrsearch.py]: Is the query generated here valid for Solr? This code looks unchanged from cloudsearch which has a fairly specific syntax.

File: r2/example.ini [hunk 3]
```diff
@@ -361,7 +366,22 @@ CLOUDSEARCH_DOC_API =
 CLOUDSEARCH_SUBREDDIT_SEARCH_API =
 # endpoint for subreddit upload
 CLOUDSEARCH_SUBREDDIT_DOC_API =
-
+#
+# endpoint for link search 
+SOLR_SEARCH_HOST = 127.0.0.1
```
Comment (kemitche) [inline r2/example.ini]: I'm a wary of adding unused config that isn't enabled by install-reddit.sh. My preference would be to leave these values unset in example.ini and document them appropriately here or in the provider implementation.

File: solr/solrconfig.xml [hunk 4]
```diff
@@ -0,0 +1,1905 @@
+<?xml version="1.0" encoding="UTF-8" ?>
```
Comment (kemitche) [inline solr/solrconfig.xml]: Is this file necessary, or would installing Solr provide a similar default file? If there aren't any reddit specific customizations in it, please omit it.

**Comment by grandpaslab** [comment]
Will do.

**Comment by kemitche** [comment]
A possibility for you to consider: if the search code can be properly provider-ized, then the Solr-specific pieces could be contained in a separate reddit-plugin on a github repo you control; that would then mean you could modify it as you see fit rather than requiring merging changes.

**Comment by kemitche** [comment]
:nail_care: (First round of feedback complete, let me know when you're ready for round 2 of review or get in touch with me on reddit if you want to further discuss/clarify anything)

**Comment by kemitche** [comment]
Closing this pull request as I believe it's been superceded by #1289
